File: 12_robustness.py
"""
12 — Robustness battery (GAN V4, V5, V6)
========================================
Project: Who Sells to the State (2026_2)

(1) Theil between-province share per year: baseline vs excluding
    security-defence buyers vs reweighted to the constant 2016-2019
    buyer-type mix (decides whether the fracture is structural or the
    militarisation of demand).
(2) Cohort survival stratified by modal client type (territorial vs not).
(3) Supplier-level exit ratio (unique winning firms / company stock).

Input:  data/processed/adjudicaciones_tipo.parquet, supplier_master.parquet,
        denominators_provincia_anio.parquet
Output: tables/tab_robustez_theil.csv, tab_robustez_supervivencia.csv,
        tab_exit_ratio_v2.csv
        figures/fig_robustez_theil.png

Usage:  python 12_robustness.py
"""
from pathlib import Path

import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd

PROJECT = Path(__file__).parent
PROC = PROJECT / "data" / "processed"
TAB = PROJECT / "tables"
FIG = PROJECT / "figures"

NEA = {"Misiones", "Corrientes", "Chaco", "Formosa"}
METRO = {"CABA", "Buenos Aires"}
ERA_OF_YEAR = {**{y: "macri" for y in range(2016, 2020)},
               **{y: "fernandez" for y in range(2020, 2024)},
               **{y: "milei" for y in range(2024, 2027)}}


def theil_between_share(firm, value="monto", group="provincia"):
    d = firm[firm[value] > 0]
    if len(d) < 2:
        return np.nan
    mu, n = d[value].mean(), len(d)
    x = d[value].to_numpy()
    total = float(np.mean((x / mu) * np.log(x / mu)))
    between = 0.0
    for _, g in d.groupby(group):
        ng, mug = len(g), g[value].mean()
        between += (ng / n) * (mug / mu) * np.log(mug / mu)
    return between / total if total else np.nan


adj = pd.read_parquet(PROC / "adjudicaciones_tipo.parquet")
adj = adj[adj["es_nuevo"] & adj["provincia"].notna() & (adj["moneda"] == "ARS")]

base_mix = (adj[adj["ejercicio"].between(2016, 2019)]["tipo_organismo"]
            .value_counts(normalize=True))
base_mix_org = (adj[adj["ejercicio"].between(2016, 2019)]["organismo"]
                .value_counts(normalize=True))

rows = []
for year, g in adj.groupby("ejercicio"):
    # baseline
    firm = g.groupby(["cuit", "provincia"], as_index=False)["monto"].sum()
    rows.append({"ejercicio": year, "variante": "baseline",
                 "share_between": theil_between_share(firm)})
    # excluding security-defence
    g2 = g[g["tipo_organismo"] != "seguridad_defensa"]
    firm2 = g2.groupby(["cuit", "provincia"], as_index=False)["monto"].sum()
    rows.append({"ejercicio": year, "variante": "sin_seguridad",
                 "share_between": theil_between_share(firm2)})
    # constant buyer mix (reweight amounts to 2016-2019 type shares)
    yr_mix = g["tipo_organismo"].value_counts(normalize=True)
    w = g["tipo_organismo"].map(base_mix / yr_mix).fillna(0)
    g3 = g.assign(monto_w=g["monto"] * w)
    firm3 = (g3.groupby(["cuit", "provincia"], as_index=False)["monto_w"].sum()
             .rename(columns={"monto_w": "monto"}))
    rows.append({"ejercicio": year, "variante": "mix_constante",
                 "share_between": theil_between_share(firm3)})
    # finer variant: constant mix at organism level (162 buyers)
    yr_org = g["organismo"].value_counts(normalize=True)
    w4 = g["organismo"].map(base_mix_org / yr_org).fillna(0)
    firm4 = (g.assign(monto_w=g["monto"] * w4)
             .groupby(["cuit", "provincia"], as_index=False)["monto_w"].sum()
             .rename(columns={"monto_w": "monto"}))
    rows.append({"ejercicio": year, "variante": "mix_constante_organismo",
                 "share_between": theil_between_share(firm4)})
rt = pd.DataFrame(rows)
rt.to_csv(TAB / "tab_robustez_theil.csv", index=False)

fig, ax = plt.subplots(figsize=(8, 4.5))
for var, style in [("baseline", "o-"), ("sin_seguridad", "s--"),
                   ("mix_constante", "^:")]:
    d = rt[(rt["variante"] == var) & rt["ejercicio"].between(2017, 2025)]
    ax.plot(d["ejercicio"], 100 * d["share_between"], style, label=var)
for x in [2019.5, 2023.5]:
    ax.axvline(x, color="grey", lw=0.8, ls=":")
ax.set_ylabel("Theil between-provincia (% del total)")
ax.set_title("Robustez de la fractura territorial (montos ARS, 2017-2025)")
ax.legend()
fig.tight_layout()
fig.savefig(FIG / "fig_robustez_theil.png", dpi=200)
plt.close(fig)

# ── (2) Survival stratified by client type ───────────────────────────────────
# Territorial-client position measured AT COHORT ENTRY (first-year awards
# only) to avoid future-data leakage (external review 2.6): a supplier must
# not be classified by clients it acquired in later eras.
first_yr = adj.groupby("cuit")["ejercicio"].transform("min")
entry_awards = adj[adj["ejercicio"] == first_yr]
cliente_modal = (entry_awards.groupby("cuit")
                 .agg(territorial=("anclaje_territorial", "mean"),
                      provincia=("provincia", "first"),
                      entrada=("ejercicio", "first")))
cliente_modal["cliente_territorial"] = cliente_modal["territorial"] > 0.5
cliente_modal["region"] = np.select(
    [cliente_modal["provincia"].isin(NEA), cliente_modal["provincia"].isin(METRO)],
    ["NEA", "Metro"], default="Resto")
cliente_modal["cohorte"] = cliente_modal["entrada"].map(ERA_OF_YEAR)

act_era = (adj.assign(era=adj["ejercicio"].map(ERA_OF_YEAR))
           .groupby(["cuit", "era"]).size().unstack().notna())
rows = []
for (reg, terr), grp in cliente_modal[cliente_modal["cohorte"] == "macri"] \
        .groupby(["region", "cliente_territorial"]):
    surv_f = act_era.reindex(grp.index)["fernandez"].fillna(False).mean()
    surv_m = act_era.reindex(grp.index)["milei"].fillna(False).mean()
    rows.append({"region": reg, "cliente_territorial": terr, "n": len(grp),
                 "surv_fernandez": round(surv_f, 3), "surv_milei": round(surv_m, 3)})
rs = pd.DataFrame(rows)
rs.to_csv(TAB / "tab_robustez_supervivencia.csv", index=False)

# ── (3) Supplier-level exit ratio ────────────────────────────────────────────
den = pd.read_parquet(PROC / "denominators_provincia_anio.parquet")
master = pd.read_parquet(PROC / "supplier_master.parquet")
stock = den[den["ejercicio"] == 2025].set_index("provincia")["stock_sociedades"]
soc_cuits = set(master.loc[master["personeria"].notna() &
                           (master["personeria"] != "Persona Fisica"), "cuit"])
firms = adj.groupby("provincia")["cuit"].nunique().rename("firmas_ganadoras")
firms_soc = (adj[adj["cuit"].isin(soc_cuits)]
             .groupby("provincia")["cuit"].nunique().rename("sociedades_ganadoras"))
ex = pd.DataFrame({"firmas_ganadoras": firms, "sociedades_ganadoras": firms_soc,
                   "stock": stock})
for col, num in [("exit_ratio_firmas", "firmas_ganadoras"),
                 ("exit_ratio_soc", "sociedades_ganadoras")]:
    ex[col] = (ex[num] / ex[num].sum()) / (ex["stock"] / ex["stock"].sum())
ex["nea"] = ex.index.isin(NEA)
ex.round(4).to_csv(TAB / "tab_exit_ratio_v2.csv")

print("Theil between-share (%): baseline vs robustez")
print(rt.pivot_table(index="ejercicio", columns="variante", values="share_between")
      .mul(100).round(1).to_string())
print("\nSupervivencia cohorte Macri por región x cliente territorial:")
print(rs.to_string(index=False))
print("\nExit ratio a nivel firma (NEA + referencia):")
sel = sorted(NEA) + ["CABA", "Buenos Aires", "Córdoba"]
print(ex.loc[sel, ["firmas_ganadoras", "sociedades_ganadoras",
                   "exit_ratio_firmas", "exit_ratio_soc"]].round(3).to_string())
print("OK -> tables/, figures/")

File: 12_robustness.R
# 12 — Robustness battery (canonical R port of 12_robustness.py).
# Run: Rscript 12_robustness.R

source("theme_house.R", chdir = TRUE)

ERA_OF_YEAR <- function(y) ifelse(y <= 2019, "macri",
                                  ifelse(y <= 2023, "fernandez", "milei"))

theil_between_share <- function(firm, value = "monto", group = "provincia") {
  d <- firm[firm[[value]] > 0 & !is.na(firm[[value]]), ]
  if (nrow(d) < 2) return(NA_real_)
  mu <- mean(d[[value]]); n <- nrow(d)
  x <- d[[value]]
  total <- mean((x / mu) * log(x / mu))
  parts <- d |> group_by(.data[[group]]) |>
    summarise(ng = n(), mug = mean(.data[[value]]), .groups = "drop")
  between <- sum((parts$ng / n) * (parts$mug / mu) * log(parts$mug / mu))
  between / total
}

adj <- read_parquet(file.path(DIR_PROC, "adjudicaciones_tipo.parquet")) |>
  filter(es_nuevo, !is.na(provincia), moneda == "ARS")

# the 2016-2019 baseline predates the arrival of the armed forces on the
# platform (see 20_platform_stability.R), so the mix it fixes is the mix of a
# different state; the 2019 baseline is the first one taken on the full set of
# buyers, and both are reported
base_mix <- adj |> filter(between(ejercicio, 2016, 2019)) |>
  count(tipo_organismo) |> mutate(share = n / sum(n)) |>
  select(tipo_organismo, share_base = share)
base_mix_org <- adj |> filter(between(ejercicio, 2016, 2019)) |>
  count(organismo) |> mutate(share = n / sum(n)) |>
  select(organismo, share_base_org = share)
base_mix_19 <- adj |> filter(ejercicio == WIN0) |>
  count(tipo_organismo) |> mutate(share = n / sum(n)) |>
  select(tipo_organismo, share_base_19 = share)

rt <- bind_rows(lapply(sort(unique(adj$ejercicio)), function(yr) {
  g <- adj |> filter(ejercicio == yr)
  firm <- g |> group_by(cuit, provincia) |>
    summarise(monto = sum(monto, na.rm = TRUE), .groups = "drop")
  g2 <- g |> filter(tipo_organismo != "seguridad_defensa")
  firm2 <- g2 |> group_by(cuit, provincia) |>
    summarise(monto = sum(monto, na.rm = TRUE), .groups = "drop")
  yr_mix <- g |> count(tipo_organismo) |> mutate(share = n / sum(n))
  w <- g |> left_join(yr_mix |> select(tipo_organismo, share_yr = share),
                      by = "tipo_organismo") |>
    left_join(base_mix, by = "tipo_organismo") |>
    mutate(w = coalesce(share_base / share_yr, 0),
           monto_w = monto * w)
  firm3 <- w |> group_by(cuit, provincia) |>
    summarise(monto = sum(monto_w, na.rm = TRUE), .groups = "drop")
  # finer variant: constant mix at organism level (162 buyers)
  yr_org <- g |> count(organismo) |> mutate(share_yr_org = n / sum(n)) |>
    select(organismo, share_yr_org)
  w4 <- g |> left_join(yr_org, by = "organismo") |>
    left_join(base_mix_org, by = "organismo") |>
    mutate(monto_w = monto * coalesce(share_base_org / share_yr_org, 0))
  firm4 <- w4 |> group_by(cuit, provincia) |>
    summarise(monto = sum(monto_w, na.rm = TRUE), .groups = "drop")
  w5 <- g |> left_join(yr_mix |> select(tipo_organismo, share_yr = share),
                       by = "tipo_organismo") |>
    left_join(base_mix_19, by = "tipo_organismo") |>
    mutate(monto_w = monto * coalesce(share_base_19 / share_yr, 0))
  firm5 <- w5 |> group_by(cuit, provincia) |>
    summarise(monto = sum(monto_w, na.rm = TRUE), .groups = "drop")
  tibble(ejercicio = yr,
         variante = c("baseline", "sin_seguridad", "mix_constante",
                      "mix_constante_organismo", "mix_constante_2019"),
         share_between = c(theil_between_share(firm),
                           theil_between_share(firm2),
                           theil_between_share(firm3),
                           theil_between_share(firm4),
                           theil_between_share(firm5)))
}))
write_tab(rt, "tab_robustez_theil.csv")

d <- rt |> filter(between(ejercicio, 2017, 2025))
p <- ggplot(d, aes(ejercicio, 100 * share_between, colour = variante,
                   shape = variante, linetype = variante)) +
  geom_line() + geom_point() +
  geom_vline(xintercept = c(2019.5, 2023.5), linetype = "dotted",
             colour = "grey50", linewidth = 0.4) +
  scale_colour_manual(values = c(baseline = "#333333",
                                 sin_seguridad = "#d62728",
                                 mix_constante = "#1f77b4")) +
  labs(x = NULL, y = "Theil between-provincia (% del total)",
       title = "Robustez de la fractura territorial (montos ARS, 2017-2025)") +
  theme_house()
save_fig(p, "fig_robustez_theil.png")

# ── Survival stratified by client type ───────────────────────────────────────
# Territorial-client position measured AT COHORT ENTRY (first-year awards
# only) to avoid future-data leakage (external review 2.6).
cliente <- adj |> group_by(cuit) |>
  filter(ejercicio == min(ejercicio)) |>
  summarise(territorial = mean(anclaje_territorial),
            provincia = first(provincia),
            entrada = first(ejercicio), .groups = "drop") |>
  mutate(cliente_territorial = territorial > 0.5,
         region = region_of(provincia),
         cohorte = ERA_OF_YEAR(entrada))

act_era <- adj |> mutate(era = ERA_OF_YEAR(ejercicio)) |>
  distinct(cuit, era) |> mutate(v = TRUE) |>
  pivot_wider(names_from = era, values_from = v, values_fill = FALSE)

rs <- cliente |> filter(cohorte == "macri") |>
  left_join(act_era, by = "cuit") |>
  group_by(region, cliente_territorial) |>
  summarise(n = n(),
            surv_fernandez = round(mean(coalesce(fernandez, FALSE)), 3),
            surv_milei = round(mean(coalesce(milei, FALSE)), 3),
            .groups = "drop")
write_tab(rs, "tab_robustez_supervivencia.csv")

# ── Supplier-level exit ratio ────────────────────────────────────────────────
den <- read_parquet(file.path(DIR_PROC, "denominators_provincia_anio.parquet"))
master <- read_parquet(file.path(DIR_PROC, "supplier_master.parquet"))
stock <- den |> filter(ejercicio == 2025) |> select(provincia, stock = stock_sociedades)
soc_cuits <- master |> filter(!is.na(personeria),
                              personeria != "Persona Fisica") |> pull(cuit)
ex <- adj |> group_by(provincia) |>
  summarise(firmas_ganadoras = n_distinct(cuit),
            sociedades_ganadoras = n_distinct(cuit[cuit %in% soc_cuits]),
            .groups = "drop") |>
  left_join(stock, by = "provincia") |>
  mutate(exit_ratio_firmas = (firmas_ganadoras / sum(firmas_ganadoras)) /
           (stock / sum(stock, na.rm = TRUE)),
         exit_ratio_soc = (sociedades_ganadoras / sum(sociedades_ganadoras)) /
           (stock / sum(stock, na.rm = TRUE)),
         nea = provincia %in% NEA)
write_tab(ex |> mutate(across(where(is.numeric), ~round(.x, 4))),
          "tab_exit_ratio_v2.csv")

cat("Theil between-share (%):\n")
print(rt |> mutate(share_between = round(100 * share_between, 1)) |>
        pivot_wider(names_from = variante, values_from = share_between), n = 12)
cat("\nSupervivencia cohorte Macri por región x cliente territorial:\n")
print(rs)
cat("\nExit ratio a nivel firma (NEA + referencia):\n")
print(ex |> filter(provincia %in% c(sort(NEA), "CABA", "Buenos Aires",
                                    "Córdoba")) |>
        select(provincia, firmas_ganadoras, sociedades_ganadoras,
               exit_ratio_firmas, exit_ratio_soc) |>
        mutate(across(where(is.numeric), ~round(.x, 3))))
